Smooth Transition Fit Function for Muons

sf1919 opened the pull request:
**Description of work.**

Added a new Smooth Transition Fit function for Muon modelling and associated documentation.

**To test:**
1. Enable the model analysis tab and plot (see #32572 for details on how to do this).
2. Open the Muon Analysis GUI and select instrument ARGUS.
2. Load runs 73404,5,8,9,12,13,16.
3. Go to the Fitting Tab and add a FLat Background anf set A0 to -0.02.
4. Add a GausOsc function with the following settings: A=0.02, Sigma=0.01, Frequency=0.3, Phi=3.3, constraints=(0<A<1).
5. Go to the sequential fitting tab and select sequentially fit all
6. Go to the Results Tab and select sample_temp from the list of log values and then click Output Results
8. Open the Model Fitting Tab and change x to sample_temp and y to f0.A0
9. In the ADS double click the workspace ```ResultsTable;f0.A0 vs sample_temp``` (it is grouped under ```ResultsTable; Parameter Combinations```) so that it will open as a plot
10. Click Fit and add a Smooth Transition Fit Function
11. You will need to change the Midpoint to 117 before running a fit
12. Check that the fit is sensible.



Fixes #32580 .

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

sf1919 committed:
Created functions and test files

Created header and source files for Smooth Transition FitFunction
Created tests for new Fit Function
Updated CMake list

sf1919 committed:
Added documentation

Added release note and documentation for new SmoothTransition fir function
Tidied up SmoothTransition header file with correct parameters

sf1919 committed:
Updated mathematical functions

Used a minus sign in the original formula incorrectly. Amended source file as required.

Ref #32580

sf1919 committed:
Updated SmoothTransitionTest

Updated formula for SmoothTransition function to match source file.
Added a test for the Jacobians

re #32580

sf1919 committed:
Updated parameter defaults and maths

Updated the default parameters and differentials in the source file
Corrected the jacobians in the tests
Corrected the formula in the header

re #32580

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/SmoothTransition.cpp:
```diff
@@ -0,0 +1,65 @@
+
+  for (size_t i = 0; i < nData; i++) {
+    double expFunc = exp((xValues[i] - midpoint) / gr);
+    double expFuncsq = pow((expFunc + 1), 2);
```
this is the denominator squared as it has the plus 1

AnthonyLim23 commented on docs/source/fitting/fitfunctions/SmoothTransition.rst:
```diff
@@ -0,0 +1,41 @@
+
+A Smooth Transition fit function is defined as:
+
+.. math:: y = A2\frac{A1-A2}{e^{-\frac{x-M}{G_R}}+1}
```
```suggestion
.. math:: y = A2+\frac{A1-A2}{e^{-\frac{x-M}{G_R}}+1}
```

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/SmoothTransition.cpp:
```diff
@@ -0,0 +1,65 @@
+  const double gr = getParameter("GrowthRate");
+
+  for (size_t i = 0; i < nData; i++) {
+    out[i] = a2 + (a1 - a2) / (exp((xValues[i] - midpoint) / gr) + 1);
```
I have had a look into this and the descriptions for a1 and a2 are the wrong way around. Also instead of saying flat background and maximum it would be better to say:


a1: the limit of the function as x tends to zero
a2: the limit of the function as x tends to infinity 

you can see these by placing considering the limits (I will do a talk about it next week). Basically the exponential goes to zero when x is small and becomes large as x becomes larger. Then you can say these terms are much larger/smaller than the others so lets ignore them

sf1919 committed:
Changes to source, header and documentation based on review

Updated parameter information in header and documentation files.
Renamed variable to denominatorSq in source file.
Added new source to documentation.

re #32580

AnthonyLim23 commented on Framework/CurveFitting/src/Functions/SmoothTransition.cpp:
```diff
@@ -23,7 +23,7 @@ using namespace API;
 
 void SmoothTransition::init() {
-  declareParameter("A1", 0.00, "maximum curve");
+  declareParameter("A1", 0.0, "maximum curve");
```
these descriptions dont match the docs, could you please update them?

sf1919 committed:
Updated paramtere descriptions in source file

re #32580

AnthonyLim23 commented:
I will need to delay testing this as the ARGUS files are missing on the archive. I think the serve/hard drive is down again

AnthonyLim23 reviewed (dismissed):
Looks good and give a very nice fit to the data

sf1919 commented:
Looks like the folder where the test helper is held has had a name change so I'll need to update the test in this before it can be merged in otherwise there will be failing tests on the main branch.

sf1919 committed:
Change MantidTestHelpers to MantidFrameworkTestHelpers in test

re #32580

sf1919 commented:
@mantidproject/gatekeepers Please can this be looked at soon? It's likely to cause merge conflicts for my other Fit Function PRs so they can't be reviewed until this one is in.
